Modules/posixmodule.c

```diff
@@ -14585,15 +14614,22 @@ DirEntry_fetch_stat(PyObject *module, DirEntry *self, int follow_symlinks)
         }
         Py_END_ALLOW_THREADS
     }
+
+    if (result != 0) {
+        path_object_error(self->path);
+#if defined(MS_WINDOWS)
+        PyMem_Free(path);
+#else
+        Py_DECREF(ub);
+#endif
```

Comment:
i dislike the fact that you have to copy/paste the cleanup code. maybe save/restor errno?

Reply:
Done.